### PR TITLE
network: Add grpc method to add static arp neighbors

### PR DESCRIPTION
**Pull request by amshinde**
Add grpc API for adding arp neighbours for a network
interface. These are expected to be static arp entries
sent by the runtime.

Fixes: #783

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

### CODE REVIEW

File: pkg/types/types.proto [hunk 1]
```diff
@@ -46,3 +46,12 @@ message Route {
 	string source = 4;
 	uint32 scope = 5;
 }
+
+message ARPNeighbor {
+	IPAddress toIPAddress = 1;
+	string device = 2;
+	string lladdr = 3;
+	int32 state = 4;
+	int32 flags = 6;
```
Comment (devimc) [inline pkg/types/types.proto:55]: and where is 5 ?

Comment (amshinde) [reply pkg/types/types.proto:55]: uh oh, missed this while deleting a field, will update.

**Comment by devimc** [comment]
CI not happy

```
protocols/mockserver/mockserver.go:54:38: cannot use mock (variable of type *mockServer) as grpc.AgentServiceServer value in argument to pb.RegisterAgentServiceServer: missing method AddARPNeighbors (typecheck)

	pb.RegisterAgentServiceServer(serv, mock)
```

File: network.go [hunk 2]
```diff
@@ -625,6 +626,58 @@ func (s *sandbox) updateRoute(netHandle *netlink.Handle, route *types.Route, add
 	return nil
 }
 
+// addARPNeighbors will take neighbors and add them to ARP entries.
+// This is used for mirror any static arp entries created by the network
```
Comment (jodh-intel) [inline network.go:630]: Nit: "is used to mirror" ?

File: network.go [hunk 3]
```diff
@@ -625,6 +626,58 @@ func (s *sandbox) updateRoute(netHandle *netlink.Handle, route *types.Route, add
...
+	}
+
+	for _, neighbor := range requestedNeighbors.ARPNeighbors {
+		// Find link index from route's device name.
+		link, err := netHandle.LinkByName(neighbor.Device)
+		if err != nil {
+			return grpcStatus.Errorf(codes.Internal, "Could not find link from device %s: %v", neighbor.Device, err)
+		}
+
+		toIP := net.ParseIP(neighbor.ToIPAddress.Address)
```
Comment (jodh-intel) [inline network.go:653]: `net.ParseIP()` returns `nil` on error so shouldn't we check that?

File: grpc.go [hunk 4]
```diff
@@ -1571,6 +1571,10 @@ func (a *agentGRPC) ListRoutes(ctx context.Context, req *pb.ListRoutesRequest) (
 	return a.sandbox.listRoutes(nil)
 }
 
+func (a *agentGRPC) AddARPNeighbors(ctx context.Context, req *pb.AddARPNeighborsRequest) (*gpb.Empty, error) {
```
Comment (jodh-intel) [inline grpc.go:1574]: This feature will also need to be added to the new rust agent.

/cc @lifupan.

Comment (lifupan) [reply grpc.go:1574]: @jodh-intel 
 I had opened an issue to track it on rust-agent repo:  https://github.com/kata-containers/kata-containers/issues/218

Comment (jodh-intel) [reply grpc.go:1574]: Thanks @lifupan !

File: network.go (hunk 2 above)
Comment (amshinde) [reply network.go]: @jodh-intel fixed

File: network.go (hunk 3 above)
Comment (amshinde) [reply network.go:653]: My initial thought was netlink would error out in that case. I have added an explicit check for it now.

**Comment by amshinde** [comment]
Seeing this in the travis logs:
```
INFO: Running 'static_check_license_headers' function
INFO: Checking for SPDX license headers
ERROR: Required license identifier ('SPDX-License-Identifier: Apache-2.0') missing from following files:
protocols/grpc/healthpb_test.go
protocols/grpc/ocipb_test.go
```
@GabyCT @chavafg Has something changed recently?

**Comment by jodh-intel** [comment]
Travis caught a problem:

```
INFO: Checking for SPDX license headers
ERROR: Required license identifier ('SPDX-License-Identifier: Apache-2.0') missing from following files:

protocols/grpc/healthpb_test.go
protocols/grpc/ocipb_test.go
```

**Comment by jodh-intel** [comment]
Doh. Snap ;) Yes, we need to add an exclude to the static checker for these. I've raised https://github.com/kata-containers/tests/pull/2543 to fix this.

**Comment by chavafg** [comment]
with @jodh-intel fix, travis now passed :)

**Pull request merged**